File: hw2/java/Chef.java
public class Chef extends Thread {

    private String name;
    private Kitchen kitchen;
    private int completedOrders = 0;
    private volatile Order job = null;
    private volatile boolean running = true;

    public Chef(Kitchen kitchen, String name) {
        this.kitchen = kitchen;
        this.name = name;
    }

    public synchronized boolean assignOrder(Order order) {
        if (this.job != null) {
            return false;
        }

        Logger.log("Order: " + order.meal() + " assigned to chef " + this.name);
        this.job = order;
        return true;
    }

    private void waitForOrder() {
        while (this.job == null) {
            Thread.yield();
        }
    }

    private void cook() {
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return;
        }

        System.out.println(
            "🍳 Chef: " +
            this.name +
            " cooked " +
            this.job.meal() +
            " for philosopher " +
            this.job.orderedBy()
        );
    }

    private void putOnKitchenCounterAndIncrementOrderCounter() {
        this.kitchen.addToKitchenOutQueue(this.job);
        this.job = null;
        this.completedOrders++;
    }

    private void takeCoffeeBreakIfItsTime() {
        if (this.completedOrders % 4 == 0 && this.completedOrders > 0) {
            System.out.println("☕ Chef coffee break");
            try {
                Thread.sleep(2000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return;
            }
        }
    }

    public boolean isBusy() {
        return this.job != null;
    }

    public void shutdown() {
        this.running = false;
        this.interrupt();
    }

    public String name() {
        return this.name;
    }

    @Override
    public void run() {
        while (this.running && !Thread.currentThread().isInterrupted()) {
            this.waitForOrder();
            this.cook();
            this.putOnKitchenCounterAndIncrementOrderCounter();
            this.takeCoffeeBreakIfItsTime();
        }

        Logger.log(this.name + "leaving for the day");
    }
}

File: hw2/go/chef.go
package main

import "time"

type Chef struct {
	name         string
	restaurant   *Restaurant
	orderChan    chan *Order
	cookedOrders int
	dayOver      chan struct{}
}

func newChef(name string, restaurant *Restaurant) *Chef {
	return &Chef{
		name:       name,
		restaurant: restaurant,
		orderChan:  make(chan *Order, 1),
		dayOver:    make(chan struct{}, 1),
	}
}

func (c *Chef) waitForOrder() *Order {
	logf("%s is waiting for order", c.name)
	return <-c.orderChan
}

func (c *Chef) cook(order *Order) {
	logf("%s is cooking order %s", c.name, order.name())
	time.Sleep(time.Second)
}

func (c *Chef) putOrderOnKitchenCounter(order *Order) {
	logf("%s finished cooking order %s and puts it on kitchen counter", c.name, order.name())
	c.restaurant.putOrderOnKitchenCounter(order)
}

func (c *Chef) takeCoffeeBreakIfItsTime() {
	if c.cookedOrders != 0 && c.cookedOrders%4 == 0 {
		logf("%s takes a coffee break", c.name)
		time.Sleep(time.Minute)
	}
}

func (c *Chef) terminate() {
	c.dayOver <- struct{}{}
}

func (c *Chef) run() {
	for {
		select {
		case <-c.dayOver:
			logf("%s is closing for the day", c.name)
			return
		case order := <-c.orderChan:
			c.cook(order)
			c.putOrderOnKitchenCounter(order)
			c.takeCoffeeBreakIfItsTime()
		}
	}
}
